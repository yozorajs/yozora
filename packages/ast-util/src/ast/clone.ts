import type { IYastNode, IYastParent, Root } from '@yozora/ast'

/**
 * Shallow clone ast until the match reaches the termination condition.
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `endCondition` function.
 *
 * @param root
 * @param endCondition
 * @returns
 */
export function shallowCloneAst(
  root: Root,
  endCondition: (node: IYastNode, parent: IYastParent, childIndex: number) => boolean,
): Root {
  const clone = (u: IYastParent): IYastParent => {
    const nextChildren = []
    const { children } = u
    for (let i = 0; i < children.length; ++i) {
      const v = children[i] as IYastParent
      if (endCondition(v, u, i)) break

      const nextChild = v.children == null ? v : clone(v)
      nextChildren.push(nextChild)
    }
    return { ...u, children: nextChildren }
  }
  return clone(root) as Root
}
