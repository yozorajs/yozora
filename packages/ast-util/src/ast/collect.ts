import type { Root, YastNode, YastNodeType, YastParent } from '@yozora/ast'
import type { NodeMatcher } from './util'
import { createNodeMatcher } from './util'

/**
 * Collect nodes of the specified type through pre-order traversal.
 *
 * Note that if the matched node has other matching descendant nodes, these
 * descendant nodes will not be collected repeatedly, that is, only the
 * currently matched node will be collected, and will not enter the subtree for
 * collection afterwards.
 *
 * @param root
 * @param aimTypesOrNodeMatcher
 */
export function collectNodes<T extends YastNodeType, O extends YastNode<T>>(
  root: Root,
  aimTypesOrNodeMatcher: ReadonlyArray<YastNodeType> | NodeMatcher | null,
): O[] {
  const isMatched: NodeMatcher = createNodeMatcher(aimTypesOrNodeMatcher)

  const nodes: O[] = []
  const collect = (u: YastParent): void => {
    const { children } = u
    if (children != null) {
      for (const v of children) collect(v as YastParent)
    }
    if (isMatched(u)) nodes.push(u as unknown as O)
  }
  collect(root)
  return nodes
}
