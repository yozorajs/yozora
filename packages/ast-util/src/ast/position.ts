import type { Node, Parent, Root } from '@yozora/ast'

export function removePositions(immutableAst: Readonly<Root>): Root {
  function remove(node: Node): Node {
    const { position, children, ...nextNode } = node as Parent
    for (const key of Object.keys(nextNode)) {
      nextNode[key] = Array.isArray(nextNode[key]) ? nextNode[key].map(remove) : nextNode[key]
    }
    if (children) {
      ;(nextNode as Parent).children = children.map(remove)
    }
    return nextNode
  }
  const tidyAst = remove(immutableAst) as Root
  return tidyAst
}
