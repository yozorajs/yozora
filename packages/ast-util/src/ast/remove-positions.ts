import type { IRoot, IYastNode, IYastParent } from '@yozora/ast'

export function removePositions(immutableAst: Readonly<IRoot>): IRoot {
  function remove(node: IYastNode): IYastNode {
    const { position, children, ...nextNode } = node as IYastParent
    if (children) {
      ;(nextNode as IYastParent).children = children.map(remove)
    }
    return nextNode
  }
  const tidyAst = remove(immutableAst) as IRoot
  return tidyAst
}
