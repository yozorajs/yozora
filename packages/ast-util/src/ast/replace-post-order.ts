import type { Node, NodeType, Parent, Root } from '@yozora/ast'
import type { INodeMatcher } from './util'
import { createNodeMatcher, createShallowNodeCollector } from './util'

/**
 * Traverse AST and replace nodes in post-order.
 *
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `replace` function as the first parameter.
 *
 * 需要注意的是，尽管此函数不会直接操作原AST，但它会尽可能地复用未改变的节点；
 * 此外它依赖于 replace 函数将传入的节点视作不可修改的，所以在使用时，需要注意不
 * 要在 replace 中直接修改节点，如果需要修改，应该返回一个新的节点（节点属性可以
 * 复用）
 *
 * It should be noted that although this function does not directly manipulate
 * the original AST, it will reuse unchanged nodes as much as possible. In
 * additional, it relies on the passed-in `replace` function to treat the passed
 * node as immutable, so when using this function, you need to be careful not
 * to modify the node directly in `replace` callback. If you need to perform
 * some changes, you should return a new node (node attributes can be reused).
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param replace
 */
export function shallowMutateAstInPostorder(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<NodeType> | INodeMatcher | null,
  replace: (
    immutableNode: Readonly<Node>,
    immutableParent: Readonly<Parent>,
    childIndex: number,
  ) => Node | Node[] | null,
): Readonly<Root> {
  const isMatched: INodeMatcher = createNodeMatcher(aimTypesOrNodeMatcher)

  const traverse = (children: ReadonlyArray<Node>, parent: Readonly<Parent>): Node => {
    // Recursively processing the descendant nodes in post-order traverse.
    const collector0 = createShallowNodeCollector(children as Node[])
    for (let i = 0; i < children.length; ++i) {
      const child = children[i] as Parent
      const subChildren: ReadonlyArray<Node> = child.children

      // Whether to process the subtree recursively.
      const nextChild = subChildren && subChildren.length > 0 ? traverse(subChildren, child) : child
      collector0.add(nextChild, child, i)
    }

    // Processing current layer of nodes.
    const nextChildren: Node[] = collector0.collect()
    const collector1 = createShallowNodeCollector(nextChildren)
    for (let i = 0; i < nextChildren.length; ++i) {
      const child = nextChildren[i]
      const nextChild = isMatched(child) ? replace(child, parent, i) : child
      collector1.add(nextChild, child, i)
    }

    const finalChildren: Node[] = collector1.collect()
    const result: Node =
      finalChildren === children ? parent : { ...parent, children: finalChildren }
    return result
  }
  return traverse(immutableRoot.children, immutableRoot) as Root
}
