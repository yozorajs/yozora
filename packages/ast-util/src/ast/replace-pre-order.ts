import type { Node, NodeType, Parent, Root } from '@yozora/ast'
import type { INodeMatcher } from './util'
import { createNodeMatcher, createShallowNodeCollector } from './util'

/**
 * Traverse AST and replace nodes in pre-order.
 *
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `replace` function as the first parameter.
 *
 * 需要注意的是，尽管此函数不会直接操作原AST，但它会尽可能地复用未改变的节点；
 * 此外它依赖于 replace 函数将传入的节点视作不可修改的，所以在使用时，需要注意不
 * 要在 replace 中直接修改节点，如果需要修改，应该返回一个新的节点（节点属性可以
 * 复用）
 *
 * It should be noted that although this function does not directly manipulate
 * the original AST, it will reuse unchanged nodes as much as possible. In
 * additional, it relies on the passed-in `replace` function to treat the passed
 * node as immutable, so when using this function, you need to be careful not
 * to modify the node directly in `replace` callback. If you need to perform
 * some changes, you should return a new node (node attributes can be reused).
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param replace
 */
export function shallowMutateAstInPreorder(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<NodeType> | INodeMatcher | null,
  replace: (
    immutableNode: Readonly<Node>,
    immutableParent: Readonly<Parent>,
    childIndex: number,
  ) => Node | Node[] | null,
): Readonly<Root> {
  const isMatched: INodeMatcher = createNodeMatcher(aimTypesOrNodeMatcher)

  const traverse = (children: ReadonlyArray<Node>, parent: Readonly<Parent>): Node => {
    // Processing current layer of nodes.
    const collector = createShallowNodeCollector(children as Node[])
    for (let i = 0; i < children.length; ++i) {
      const child = children[i] as Parent
      if (isMatched(child)) {
        const nextChild = replace(child, parent, i)
        collector.add(nextChild, child, i)
      } else {
        // Recursively processing the descendant nodes in pre-order traverse.
        const subChildren: ReadonlyArray<Node> = child.children

        // Whether to process the subtree recursively.
        const nextChild: Node =
          subChildren && subChildren.length > 0 ? traverse(subChildren, child) : child
        collector.add(nextChild, child, i)
      }
    }

    const finalChildren: Node[] = collector.collect()
    const result: Node =
      finalChildren === children ? parent : { ...parent, children: finalChildren }
    return result
  }
  return traverse(immutableRoot.children, immutableRoot) as Root
}
