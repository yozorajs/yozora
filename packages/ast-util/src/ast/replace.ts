import type { Root, YastNode, YastNodeType, YastParent } from '@yozora/ast'

/**
 * Traverse AST and replace nodes.
 * @param root
 * @param aimTypes
 * @param replace
 */
export function replaceAST(
  root: Root,
  aimTypes: ReadonlyArray<YastNodeType> | null,
  replace: (
    node: Readonly<YastNode>,
    parent: Readonly<YastParent>,
    childIndex: number,
  ) => YastNode | YastNode[] | void,
): void {
  const visit = (o: YastNode): void => {
    const u: YastParent = o as YastParent

    // Recursively visit.
    if (u.children != null) {
      const { children } = u
      let nextChildren: YastNode[] | null = null
      for (let i = 0; i < children.length; ++i) {
        const v = children[i]
        let replaced = false
        if (aimTypes == null || aimTypes.indexOf(v.type) > -1) {
          const newChild = replace(v, u, i)
          if (newChild != null) {
            replaced = true
            if (nextChildren == null) nextChildren = children.slice(0, i)
            if (Array.isArray(newChild)) nextChildren.push(...newChild)
            else nextChildren.push(newChild)
          }
        }

        if (!replaced) {
          visit(v)
          if (nextChildren != null) nextChildren.push(v)
        }
      }

      // Replace children if any of them has been replaced.
      if (nextChildren != null) {
        u.children = nextChildren
      }
    }
  }
  visit(root)
}
