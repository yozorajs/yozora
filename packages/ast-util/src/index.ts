export * from './ast/clone'
export * from './ast/collect'
export * from './ast/replace-post-order'
export * from './ast/replace-pre-order'
export * from './ast/search'
export * from './ast/toc'
export * from './ast/traverse'
export * from './ast/util'
export * from './definition'
export * from './excerpt'
export * from './footnote-definition'
export * from './url'
