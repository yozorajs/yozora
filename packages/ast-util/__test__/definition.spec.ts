import type { IDefinition, IRoot } from '@yozora/ast'
import { DefinitionType } from '@yozora/ast'
import { loadJSONFixture } from 'jest.setup'
import { calcDefinitionMap, calcIdentifierSet, collectDefinitions } from '../src'

describe('calcIdentifierSet', function () {
  describe('basic1', function () {
    const originalAst: Readonly<IRoot> = loadJSONFixture('basic1.ast.json')
    const ast: IRoot = loadJSONFixture('basic1.ast.json')

    test('default', function () {
      const result = calcIdentifierSet(ast, [DefinitionType])
      expect(result).toMatchSnapshot()
      expect(ast).toEqual(originalAst)
    })

    test('presetIdentifiers', function () {
      const result = calcIdentifierSet(
        ast,
        [DefinitionType],
        [
          {
            identifier: 'foooo',
            label: 'ooooof',
          },
        ],
      )
      expect(result).toMatchSnapshot()
      expect(ast).toEqual(originalAst)
    })
  })
})

describe('collectDefinitions', function () {
  test('basic1', function () {
    const originalAst: Readonly<IRoot> = loadJSONFixture('basic1.ast.json')
    const ast: IRoot = loadJSONFixture('basic1.ast.json')

    const result = collectDefinitions(ast)
    expect(result).toMatchSnapshot()
    expect(ast).toEqual(originalAst)
  })
})

describe('calcDefinitionMap', function () {
  test('basic1', function () {
    const originalAst: Readonly<IRoot> = loadJSONFixture('basic1.ast.json')
    const ast: IRoot = loadJSONFixture('basic1.ast.json')

    const { root, definitionMap } = calcDefinitionMap(ast)
    expect(root).toBe(ast)
    expect(ast).toEqual(originalAst)
    expect(definitionMap).toMatchSnapshot()
  })

  test('presetDefinitions', function () {
    const originalAst: Readonly<IRoot> = loadJSONFixture('basic1.ast.json')
    const ast: IRoot = loadJSONFixture('basic1.ast.json')
    const presetDefinitions: IDefinition[] = [
      {
        type: DefinitionType,
        identifier: 'yozora__test_identifier',
        label: 'yozora test label',
        url: '#/',
        title: 'waw',
      },
    ]

    const { root, definitionMap } = calcDefinitionMap(ast, undefined, presetDefinitions)

    expect(root).not.toBe(ast)
    expect(root).toEqual({
      ...ast,
      children: ast.children.concat(presetDefinitions),
    })
    expect(ast).toEqual(originalAst)
    expect(definitionMap).toMatchSnapshot()
  })
})
