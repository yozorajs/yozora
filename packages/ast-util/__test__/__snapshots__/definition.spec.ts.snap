// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcDefinitionMap basic1 1`] = `
{
  "bar": {
    "identifier": "bar",
    "label": "bar",
    "position": {
      "end": {
        "column": 22,
        "line": 83,
        "offset": 1295,
      },
      "start": {
        "column": 1,
        "line": 83,
        "offset": 1274,
      },
    },
    "title": "title2",
    "type": "definition",
    "url": "/bar",
  },
  "baz": {
    "identifier": "baz",
    "label": "baz",
    "position": {
      "end": {
        "column": 13,
        "line": 84,
        "offset": 1307,
      },
      "start": {
        "column": 1,
        "line": 84,
        "offset": 1295,
      },
    },
    "type": "definition",
    "url": "/waw",
  },
  "logo": {
    "identifier": "logo",
    "label": "logo",
    "position": {
      "end": {
        "column": 44,
        "line": 82,
        "offset": 1274,
      },
      "start": {
        "column": 1,
        "line": 82,
        "offset": 1231,
      },
    },
    "title": "The logo of Yozora",
    "type": "definition",
    "url": "/img/logo.png",
  },
  "url2": {
    "identifier": "url2",
    "label": "url2",
    "position": {
      "end": {
        "column": 30,
        "line": 85,
        "offset": 1336,
      },
      "start": {
        "column": 1,
        "line": 85,
        "offset": 1307,
      },
    },
    "type": "definition",
    "url": "../url//ss/s/./../sd",
  },
}
`;

exports[`calcDefinitionMap presetDefinitions 1`] = `
{
  "bar": {
    "identifier": "bar",
    "label": "bar",
    "position": {
      "end": {
        "column": 22,
        "line": 83,
        "offset": 1295,
      },
      "start": {
        "column": 1,
        "line": 83,
        "offset": 1274,
      },
    },
    "title": "title2",
    "type": "definition",
    "url": "/bar",
  },
  "baz": {
    "identifier": "baz",
    "label": "baz",
    "position": {
      "end": {
        "column": 13,
        "line": 84,
        "offset": 1307,
      },
      "start": {
        "column": 1,
        "line": 84,
        "offset": 1295,
      },
    },
    "type": "definition",
    "url": "/waw",
  },
  "logo": {
    "identifier": "logo",
    "label": "logo",
    "position": {
      "end": {
        "column": 44,
        "line": 82,
        "offset": 1274,
      },
      "start": {
        "column": 1,
        "line": 82,
        "offset": 1231,
      },
    },
    "title": "The logo of Yozora",
    "type": "definition",
    "url": "/img/logo.png",
  },
  "url2": {
    "identifier": "url2",
    "label": "url2",
    "position": {
      "end": {
        "column": 30,
        "line": 85,
        "offset": 1336,
      },
      "start": {
        "column": 1,
        "line": 85,
        "offset": 1307,
      },
    },
    "type": "definition",
    "url": "../url//ss/s/./../sd",
  },
  "yozora__test_identifier": {
    "identifier": "yozora__test_identifier",
    "label": "yozora test label",
    "title": "waw",
    "type": "definition",
    "url": "#/",
  },
}
`;

exports[`calcIdentifierSet basic1 default 1`] = `
Set {
  "logo",
  "bar",
  "baz",
  "url2",
}
`;

exports[`calcIdentifierSet basic1 presetIdentifiers 1`] = `
Set {
  "logo",
  "bar",
  "baz",
  "url2",
  "foooo",
}
`;

exports[`collectDefinitions basic1 1`] = `
[
  {
    "identifier": "logo",
    "label": "logo",
    "position": {
      "end": {
        "column": 44,
        "line": 82,
        "offset": 1274,
      },
      "start": {
        "column": 1,
        "line": 82,
        "offset": 1231,
      },
    },
    "title": "The logo of Yozora",
    "type": "definition",
    "url": "/img/logo.png",
  },
  {
    "identifier": "bar",
    "label": "bar",
    "position": {
      "end": {
        "column": 22,
        "line": 83,
        "offset": 1295,
      },
      "start": {
        "column": 1,
        "line": 83,
        "offset": 1274,
      },
    },
    "title": "title2",
    "type": "definition",
    "url": "/bar",
  },
  {
    "identifier": "baz",
    "label": "baz",
    "position": {
      "end": {
        "column": 13,
        "line": 84,
        "offset": 1307,
      },
      "start": {
        "column": 1,
        "line": 84,
        "offset": 1295,
      },
    },
    "type": "definition",
    "url": "/waw",
  },
  {
    "identifier": "url2",
    "label": "url2",
    "position": {
      "end": {
        "column": 30,
        "line": 85,
        "offset": 1336,
      },
      "start": {
        "column": 1,
        "line": 85,
        "offset": 1307,
      },
    },
    "type": "definition",
    "url": "../url//ss/s/./../sd",
  },
]
`;
